packages/cursorless-engine/src/core/commandVersionUpgrades/generateSpokenForm.ts

```diff
@@ -0,0 +1,477 @@
+  const prefix = type === "vertical" ? `${connectives.verticalRange} ` : "";
+  if (excludeAnchor && excludeActive) {
+    return prefix + connectives.rangeExclusive;
+  }
+  if (excludeAnchor) {
+    throw new NoSpokenFormError("Range exclude anchor");
+  }
+  if (excludeActive) {
+    return prefix + connectives.rangeExcludingEnd;
+  }
+  if (type === "vertical") {
+    // "slice", but could have been "slice past"
+    return connectives.verticalRange;
+  }
+  return connectives.rangeInclusive;
+}
+
+function insertionModeToSpokenForm(insertionMode: InsertionMode): string {
+  switch (insertionMode) {
+    case "to":
+      return connectives.sourceDestinationConnective;
+    case "before":
+      return connectives.before;
+    case "after":
+      return connectives.after;
+  }
+}
+
+function primitiveTargetToSpokenForm(
+  target: PartialPrimitiveTargetDescriptor,
+): RecursiveArray<string> {
+  const components: RecursiveArray<string> = [];
+  if (target.modifiers != null) {
+    components.push(target.modifiers.map(modifierToSpokenForm));
+  }
+  if (target.mark != null) {
+    components.push(markToSpokenForm(target.mark));
+  }
+  return components;
+}
+
+function modifierToSpokenForm(modifier: Modifier): RecursiveArray<string> {
+  switch (modifier.type) {
+    case "cascading":
+    case "modifyIfUntyped":
+      throw new NoSpokenFormError(`Modifier '${modifier.type}'`);
+
+    case "containingScope":
+      return [scopeTypeToSpokenForm(modifier.scopeType)];
+
+    case "everyScope":
+      return [modifiers.everyScope, scopeTypeToSpokenForm(modifier.scopeType)];
+
+    case "extendThroughStartOf":
+    case "extendThroughEndOf": {
+      const type = modifiers[modifier.type];
+      return modifier.modifiers != null
+        ? [type, modifier.modifiers.map(modifierToSpokenForm)]
+        : [type];
+    }
+
+    case "relativeScope":
+      return modifier.offset === 0
+        ? relativeScopeInclusiveToSpokenForm(modifier)
+        : relativeScopeExclusiveToSpokenForm(modifier);
+
+    case "ordinalScope": {
+      const scope = scopeTypeToSpokenForm(modifier.scopeType);
+
+      if (modifier.length === 1) {
+        if (modifier.start === -1) {
+          return [modifiersExtra.last, scope];
+        }
+        if (modifier.start === 0) {
+          return [modifiersExtra.first, scope];
+        }
+        if (modifier.start < 0) {
+          return [
+            ordinalToSpokenForm(Math.abs(modifier.start)),
+            modifiersExtra.last,
+            scope,
+          ];
+        }
+        return [ordinalToSpokenForm(modifier.start + 1), scope];
+      }
+
+      const number = numberToSpokenForm(modifier.length);
+
+      if (modifier.start === 0) {
+        return [modifiersExtra.first, number, pluralize(scope)];
+      }
+      if (modifier.start === -modifier.length) {
+        return [modifiersExtra.last, number, pluralize(scope)];
+      }
+
+      throw Error(`Modifier '${modifier.type}' is not fully implemented`);
```

Comment:
How come we don't throw `NoSpokenForm` here?

Reply:
This is more of an "this shouldn't happen and if so something in my logic has gone wrong and need to be looked over" compared to "this is expected, but we don't have a spoken form"

Reply:
But there's nothing inherently wrong with having `length > 1` and `start > 0`? And we do support it in the extension I'm pretty sure.  And we could definitely introduce a spoken form for it in the future, eg `"second three words`".  So I'd argue that the only reason we're throwing an exception here is that there is currently no spoken form for it.

Reply:
If you prefer please change it

Reply:
I guess the main reason to keep all these cases as errors is that we don't have tests for them, so if we added spoken forms for them in the future, we wouldn't notice that generation doesn't work and they just get skipped in CI.  And their talon spoken form tests would get skipped too.  So I think I'd argue to either keep them as errors or add tests for them.  I think maybe for "second two funks" and "second next two funks", I'll add tests and switch them to no spoken form, and for range modifier / mark, let's leave as error, because it's unclear we'll ever support them, and when we do, we'll notice quickly when the error gets thrown.  Make sense?

Reply:
Tbh I'm kinda tempted to capture which tests don't have proper spoken forms in their fixtures.  I'm not in love with the silent / possibly unpredictable skipping. Should be easy for us to automatically add those fields now.  Would be good to notice if they change.  wdyt?

Reply:
Fwiw that would further decouple this PR from talon test PR, because then the field is just in the test itself, so it doesn't even need to call `generateSpokenForms`.  All in all, I think the idea of using some complex logic to determine whether to skip a test feels like an anti-pattern the more I think about it

Reply:
That was kind of the purpose of keeping them as errors. If something in the future changes we will not silently as drop them.

Reply:
Ok but to be clear this is what I'm proposing:

1. In your `generateSpokenForms`, keep everything as it is except that we change `"second two funks"` and `"second next two funks"` to throw `NoSpokenFormError`
2. Record tests for`"second two funks"` and `"second next two funks"`
3. Add a field to all recorded tests that currently have no spoken form
4. Use that field to decide whether we're ok with spoken form being `null`
5. In your PR, use that field to decide whether to run Talon test 

Otherwise, we will not notice if something starts throwing `NoSpokenFormError` in the future. It will just silently be dropped.  Correct?

Reply:
I think that is fine.

Reply:
Cool I'm happy to take a crack at it. I'd also be tempted to let `NoSpokenFormError` bubble out of `generateSpokenForms` fwiw because today there's no way for the caller to know why there's no spoken form. But it's a bit unfortunate that the type system gives them no indication that they should be expecting that error. In fact, that case is not really exceptional at all, because the user may have just disabled a spoken form. So maybe let's leave it as null for now, but consider making `generateSpokenForms` return an error object that includes message if we need that message at some point in the future

Reply:
Please do
I think it's probably better to just use null for now so we get that help from the type system. Returning in an object is of course also an alternative.

Reply:
done.  last checkbox is for your talon test PR so I think we can resolve this issue here